File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
using namespace std;

void dfs(map<string, int> &m,string s,string a,int size,int start){
    if(a.size() == size){
        m[a]++;    
        return;
    }
    for(int i=start; i < s.size();i++)
        dfs(m,s,a + s[i],size,i+1);   
}

vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> answer;
    
    for(auto c:course){
        
        map<string, int> m;
        
        for(auto s: orders){
            sort(s.begin(), s.end());
            dfs(m,s,"",c,0);
        }
        
        int Max = 0;
        for(auto s:m){
            Max = max(Max,s.second);
        }
        
        if(Max<2)
            continue;
        
        for(auto s:m){
            if(s.second == Max)
                answer.push_back(s.first);
        }
        
    }
    sort(answer.begin(),answer.end());
    
    return answer;
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.io.*;
import java.util.*;
class Solution {
    static Map<String, Integer> map = new HashMap<>();
    public String[] solution(String[] orders, int[] course) {
        String[] answer = {};
        
        ArrayList<String> an = new ArrayList<>();
        for(int c : course){
            map.clear();
            for(String s : orders){
                dfs(s, c, -1, "");
            }
            int max = 0;
            for(String key : map.keySet()){
                max = Math.max(max, map.get(key));
            }
            if(max < 2) continue;
            for(String key : map.keySet()){
                if(max == map.get(key)) an.add(key);
            }
        }
        answer = new String[an.size()];
        for(int i = 0; i < an.size(); i++) answer[i] = an.get(i);
        Arrays.sort(answer, (a,b)->a.compareTo(b));
        
        return answer;
    }
    static void dfs(String order, int max, int index, String s){
        if(s.length() == max){
            char[] ch = s.toCharArray();
            Arrays.sort(ch);
            String temp = new String(ch);
            if(map.containsKey(temp)) map.put(temp, map.get(temp) + 1);
            else map.put(temp, 1);
            return;
        }
        
        for(int i = index + 1; i < order.length(); i++)
            dfs(order, max, i, s + order.charAt(i));
    }
}